### ISSUE TITLE
Dagger Android with Android Test Orchestrator casues "Test framework quit unexpectedly"

### ISSUE THREAD

**Issue by andrew-aernos** [android/architecture-samples#719]
In Dagger Android example, when changed the testOptions to use Android Test Orchestrator
```
    testOptions{
        execution 'ANDROIDX_TEST_ORCHESTRATOR' // <- add this
        unitTests{
            includeAndroidResources = true
            all {
                testLogging {
                    events 'passed', 'skipped', 'failed', 'standardOut', 'standardError'
                }
            }
        }
    }
```
then run instrumented tests, the tests will executed on the device but in android studio it will report
Test framework quit unexpectedly.

This is probably due to using the CustomTestRunner that extends AndroidJUnitRunner?

**Comment by VaradGupta23** [comment]
Yes, you're absolutely right — this issue typically stems from incompatibility between Android Test Orchestrator and a custom AndroidJUnitRunner, especially if the runner does something non-standard or doesn't work well with how Orchestrator manages test execution.

Problem Summary
You’ve configured:


testOptions {
    execution 'ANDROIDX_TEST_ORCHESTRATOR'
}
But when you run instrumented tests in Android Studio, it reports:

Test framework quit unexpectedly

And your project uses Dagger Android with a custom test runner (e.g., extending AndroidJUnitRunner for Hilt or DI setup).

 Why This Happens
Android Test Orchestrator runs each test in its own Instrumentation instance, using OrchestratedInstrumentation behind the scenes.

If your custom AndroidJUnitRunner (e.g., for setting up Dagger, Hilt, test injection, or shared test state) doesn’t handle this properly, the orchestrator might crash or terminate prematurely.

Orchestrator expects very clean isolation — no static state, custom initialization hacks, or app process leaks.

 Solution Options
1.  Ensure Proper Compatibility in Custom Runner
If you're extending AndroidJUnitRunner, make sure it calls:


super.onCreate(arguments)
early in the onCreate() method. Example:


class CustomTestRunner : AndroidJUnitRunner() {
    override fun onCreate(arguments: Bundle) {
        // Important: must call this first
        super.onCreate(arguments)
        
        // Any Dagger/Hilt or test-specific setup here
    }
}
2.  Use Hilt's Official Test Runner (if using Hilt)
If you're using Hilt, your test runner must extend:


@CustomTestApplication(MyApplication::class)
class HiltTestRunner : AndroidJUnitRunner()
You must also annotate your test classes properly and avoid global state or side effects between tests.

3.  Add Required Test Instrumentation Dependencies
Make sure you’ve added the orchestrator + runner in your build.gradle:


androidTestUtil 'androidx.test:orchestrator:1.4.2'

dependencies {
    androidTestImplementation 'androidx.test:runner:1.5.2'
    androidTestImplementation 'androidx.test:rules:1.5.0'
}
4.  Try Switching to Default Runner Temporarily
To confirm the problem is with your runner, try using the default runner:

In build.gradle:


defaultConfig {
    testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
}
If the orchestrator works with the default runner but not with your custom one — your runner likely needs cleanup.

 Optional: Debug with Logcat
Check Logcat logs for:

Process crashed while running instrumentation

java.lang.RuntimeException: Instrumentation failed

This can help pinpoint if:

Your test runner is not compatible

There are static leaks or bad initialization paths

 TL;DR Fix Checklist
 super.onCreate(arguments) in your test runner
 No retained state across tests
 Hilt/Dagger setup is isolated per test
 No shared static fields in application/test classes
Try the default AndroidJUnitRunner to confirm
